File: cpp/rna-transcription/rna_transcription.cpp
#include "rna_transcription.h"
#include <stdexcept> // 예외 처리를 위해 필요한 헤더

namespace rna_transcription
{
    // DNA 염기 서열을 RNA 염기 서열로 변환하는 함수 정의
    std::string to_rna(const std::string &dna)
    {
        std::string rna;
        rna.reserve(dna.size()); // RNA 문자열의 크기를 미리 예약하여 성능 최적화

        for (char nucleotide : dna)
        {
            rna += to_rna(nucleotide); // DNA 염기를 RNA 염기로 변환하여 RNA 문자열에 추가
        }
        return rna;
    }

    // DNA 염기를 RNA 염기로 변환하는 함수 정의((단일 문자)
    char to_rna(char dna_nucleotide)
    {
        switch (dna_nucleotide)
        {
        case 'G':
            return 'C';
        case 'C':
            return 'G';
        case 'T':
            return 'A';
        case 'A':
            return 'U';
        default:
            // 예외 처리? 잘못된 염기 문자가 들어올 경우
            throw std::invalid_argument("Invalid DNA nucleotide");
        }
    }

} // namespace rna_transcription

File: python/rna-transcription/rna_transcription.py
def to_rna(dna_strand):
    # RNA 상보적인 염기 매핑
    dna_to_rna = {'G': 'C', 'C': 'G', 'T': 'A', 'A': 'U'}

    # 입력된 DNA 문자열이 빈 문자열인 경우 빈 문자열 반환
    if dna_strand == '':
        return ''
    
    # DNA 문자열을 RNA 문자열로 변환
    rna_strand = ''
    for nucleotide in dna_strand:
        if nucleotide in dna_to_rna:
            rna_strand += dna_to_rna[nucleotide]
        else:
            # 올바르지 않은 DNA 염기가 포함되어 있을 경우 오류 처리
            return None
        
    return rna_strand